File: frontend/src/api/relations.ts
import { apiFetch } from "./client";
import type { RelationRead } from "../types/relation";
import type { JsonObject } from "../types/json";

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}

export interface RoleWrite {
  entity_id: string;
  role_type: string;
  weight?: number;
  coverage?: number;
}

export interface RelationWrite {
  source_id: string;
  kind: string;
  direction?: string;
  confidence?: number;
  scope?: JsonObject | null;
  notes?: Record<string, string>;
  roles: RoleWrite[];
  created_with_llm?: string;
}

export function listRelationsBySource(sourceId: string): Promise<RelationRead[]> {
  return apiFetch(`/relations/by-source/${sourceId}`);
}

export function listRelations(limit = 50, offset = 0): Promise<PaginatedResponse<RelationRead>> {
  return apiFetch(`/relations?limit=${limit}&offset=${offset}`);
}

export function getRelation(relationId: string): Promise<RelationRead> {
  return apiFetch(`/relations/${relationId}`);
}

export function createRelation(payload: RelationWrite): Promise<RelationRead> {
  return apiFetch("/relations/", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateRelation(id: string, payload: RelationWrite): Promise<RelationRead> {
  return apiFetch(`/relations/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteRelation(id: string): Promise<void> {
  return apiFetch(`/relations/${id}`, {
    method: "DELETE",
  });
}

File: backend/app/api/relations.py
from fastapi import APIRouter, Depends, Query
from uuid import UUID

from app.api.service_dependencies import get_relation_service
from app.schemas.relation import RelationWrite, RelationRead
from app.schemas.pagination import PaginatedResponse
from app.services.relation_service import RelationService
from app.dependencies.auth import get_current_user

router = APIRouter()


@router.post("/", response_model=RelationRead)
async def create_relation(
    payload: RelationWrite,
    service: RelationService = Depends(get_relation_service),
    user=Depends(get_current_user),  # 🔒 auth required
):
    return await service.create(payload, user_id=user.id if user else None)

@router.get("/", response_model=PaginatedResponse[RelationRead])
async def list_relations(
    limit: int = Query(50, description="Maximum number of results", ge=1, le=100),
    offset: int = Query(0, description="Number of results to skip", ge=0),
    service: RelationService = Depends(get_relation_service),
):
    return await service.list_all(limit=limit, offset=offset)

@router.get("/by-source/{source_id}", response_model=list[RelationRead])
async def list_relations_by_source(
    source_id: UUID,
    service: RelationService = Depends(get_relation_service),
):
    # Intentionally unauthenticated: public read endpoint for source-scoped relations.
    return await service.list_by_source(source_id)

@router.get("/{relation_id}", response_model=RelationRead)
async def get_relation(
    relation_id: UUID,
    service: RelationService = Depends(get_relation_service),
):
    return await service.get(relation_id)

@router.put("/{relation_id}", response_model=RelationRead)
async def update_relation(
    relation_id: UUID,
    payload: RelationWrite,
    service: RelationService = Depends(get_relation_service),
    user=Depends(get_current_user),  # 🔒 auth required
):
    return await service.update(relation_id, payload, user_id=user.id if user else None)

@router.delete("/{relation_id}", status_code=204)
async def delete_relation(
    relation_id: UUID,
    service: RelationService = Depends(get_relation_service),
    user=Depends(get_current_user),  # 🔒 auth required
):
    await service.delete(relation_id)
    return None
